File: web/js/algorithms/Dijkstra.js
/**
 * Dijkstra's Algorithm Implementation
 * Finds shortest path in weighted graph using priority queue
 */

class DijkstraAlgorithm {
    constructor(graph) {
        this.graph = graph;
    }

    /**
     * Find shortest path from start to end vertex
     * @param {string} start - Start vertex
     * @param {string} end - End vertex
     * @returns {Object} {path: string[], distance: number}
     */
    findShortestPath(start, end) {
        if (!this.graph.hasVertex(start) || !this.graph.hasVertex(end)) {
            return { path: [], distance: -1 };
        }

        // Initialize distances and visited set
        const distances = new Map();
        const previous = new Map();
        const visited = new Set();
        const pq = new PriorityQueue();

        // Initialize all distances to infinity
        for (const vertex of this.graph.getVertices()) {
            distances.set(vertex, Infinity);
        }
        distances.set(start, 0);

        // Add start vertex to priority queue
        pq.enqueue(start, 0);

        while (!pq.isEmpty()) {
            const current = pq.dequeue();

            if (visited.has(current)) continue;
            visited.add(current);

            // If we reached the end vertex, we can stop early
            if (current === end) break;

            // Check all neighbors
            const neighbors = this.graph.getNeighbors(current);
            for (const neighbor of neighbors) {
                if (visited.has(neighbor)) continue;

                const edgeWeight = this.graph.getEdgeWeight(current, neighbor);
                const newDistance = distances.get(current) + edgeWeight;

                if (newDistance < distances.get(neighbor)) {
                    distances.set(neighbor, newDistance);
                    previous.set(neighbor, current);
                    pq.enqueue(neighbor, newDistance);
                }
            }
        }

        // Check if path exists
        if (distances.get(end) === Infinity) {
            return { path: [], distance: -1 };
        }

        // Reconstruct path
        const path = this.reconstructPath(previous, start, end);
        return { path, distance: distances.get(end) };
    }

    /**
     * Find shortest distances from start vertex to all other vertices
     * @param {string} start - Start vertex
     * @returns {Map} Map of vertex to distance
     */
    findShortestDistances(start) {
        if (!this.graph.hasVertex(start)) {
            return new Map();
        }

        const distances = new Map();
        const visited = new Set();
        const pq = new PriorityQueue();

        // Initialize all distances to infinity
        for (const vertex of this.graph.getVertices()) {
            distances.set(vertex, Infinity);
        }
        distances.set(start, 0);

        pq.enqueue(start, 0);

        while (!pq.isEmpty()) {
            const current = pq.dequeue();

            if (visited.has(current)) continue;
            visited.add(current);

            const neighbors = this.graph.getNeighbors(current);
            for (const neighbor of neighbors) {
                if (visited.has(neighbor)) continue;

                const edgeWeight = this.graph.getEdgeWeight(current, neighbor);
                const newDistance = distances.get(current) + edgeWeight;

                if (newDistance < distances.get(neighbor)) {
                    distances.set(neighbor, newDistance);
                    pq.enqueue(neighbor, newDistance);
                }
            }
        }

        return distances;
    }

    /**
     * Check if a path exists between two vertices
     * @param {string} start - Start vertex
     * @param {string} end - End vertex
     * @returns {boolean} True if path exists
     */
    pathExists(start, end) {
        const result = this.findShortestPath(start, end);
        return result.path.length > 0;
    }

    /**
     * Get all vertices reachable from start vertex
     * @param {string} start - Start vertex
     * @returns {string[]} Array of reachable vertices
     */
    getReachableVertices(start) {
        const distances = this.findShortestDistances(start);
        const reachable = [];
        
        for (const [vertex, distance] of distances) {
            if (distance !== Infinity) {
                reachable.push(vertex);
            }
        }
        
        return reachable;
    }

    /**
     * Reconstruct path from previous map
     * @param {Map} previous - Map of vertex to previous vertex
     * @param {string} start - Start vertex
     * @param {string} end - End vertex
     * @returns {string[]} Path from start to end
     */
    reconstructPath(previous, start, end) {
        const path = [];
        let current = end;

        while (current !== undefined) {
            path.unshift(current);
            current = previous.get(current);
        }

        return path;
    }

    /**
     * Get algorithm complexity information
     * @returns {Object} Complexity information
     */
    getComplexityInfo() {
        return {
            timeComplexity: 'O((V + E) log V)',
            spaceComplexity: 'O(V)',
            description: 'Dijkstra\'s algorithm with binary heap priority queue'
        };
    }
}

File: cpp/src/Dijkstra.cpp
#include "../include/Dijkstra.h"
#include <limits>
#include <algorithm>
#include <functional>

Dijkstra::Dijkstra(const Graph& g) : graph(g) {
    // Constructor
}

Dijkstra::~Dijkstra() {
    // Destructor
}

std::pair<std::vector<std::string>, int> Dijkstra::findShortestPath(
    const std::string& start, 
    const std::string& end
) {
    // Validate input vertices
    if (!validateVertices(start, end)) {
        return {{}, -1};
    }
    
    // Initialize distance map with infinity
    std::unordered_map<std::string, int> distances;
    std::unordered_map<std::string, std::string> parent;
    std::unordered_set<std::string> visited;
    
    // Get all vertices
    auto vertices = graph.getVertices();
    for (const auto& vertex : vertices) {
        distances[vertex] = std::numeric_limits<int>::max();
    }
    
    // Distance to start vertex is 0
    distances[start] = 0;
    
    // Priority queue for Dijkstra's algorithm (min-heap)
    std::priority_queue<Node, std::vector<Node>, std::greater<Node>> pq;
    pq.push(Node(start, 0));
    
    while (!pq.empty()) {
        Node current = pq.top();
        pq.pop();
        
        std::string currentVertex = current.vertex;
        
        // Skip if already visited
        if (visited.find(currentVertex) != visited.end()) {
            continue;
        }
        
        visited.insert(currentVertex);
        
        // If we reached the end vertex, we can stop early
        if (currentVertex == end) {
            break;
        }
        
        // Get neighbors of current vertex
        auto neighbors = graph.getNeighbors(currentVertex);
        
        for (const auto& neighbor : neighbors) {
            if (visited.find(neighbor) == visited.end()) {
                int edgeWeight = graph.getEdgeWeight(currentVertex, neighbor);
                int newDistance = distances[currentVertex] + edgeWeight;
                
                if (newDistance < distances[neighbor]) {
                    distances[neighbor] = newDistance;
                    parent[neighbor] = currentVertex;
                    pq.push(Node(neighbor, newDistance));
                }
            }
        }
    }
    
    // Check if path exists
    if (distances[end] == std::numeric_limits<int>::max()) {
        return {{}, -1}; // No path found
    }
    
    // Reconstruct path
    std::vector<std::string> path = reconstructPath(parent, start, end);
    
    return {path, distances[end]};
}

std::unordered_map<std::string, int> Dijkstra::findShortestDistances(
    const std::string& start
) {
    std::unordered_map<std::string, int> distances;
    std::unordered_set<std::string> visited;
    
    // Get all vertices
    auto vertices = graph.getVertices();
    for (const auto& vertex : vertices) {
        distances[vertex] = std::numeric_limits<int>::max();
    }
    
    // Distance to start vertex is 0
    distances[start] = 0;
    
    // Priority queue for Dijkstra's algorithm
    std::priority_queue<Node, std::vector<Node>, std::greater<Node>> pq;
    pq.push(Node(start, 0));
    
    while (!pq.empty()) {
        Node current = pq.top();
        pq.pop();
        
        std::string currentVertex = current.vertex;
        
        // Skip if already visited
        if (visited.find(currentVertex) != visited.end()) {
            continue;
        }
        
        visited.insert(currentVertex);
        
        // Get neighbors of current vertex
        auto neighbors = graph.getNeighbors(currentVertex);
        
        for (const auto& neighbor : neighbors) {
            if (visited.find(neighbor) == visited.end()) {
                int edgeWeight = graph.getEdgeWeight(currentVertex, neighbor);
                int newDistance = distances[currentVertex] + edgeWeight;
                
                if (newDistance < distances[neighbor]) {
                    distances[neighbor] = newDistance;
                    pq.push(Node(neighbor, newDistance));
                }
            }
        }
    }
    
    return distances;
}

bool Dijkstra::pathExists(const std::string& start, const std::string& end) {
    auto result = findShortestPath(start, end);
    return !result.first.empty();
}

std::vector<std::string> Dijkstra::getReachableVertices(const std::string& start) {
    auto distances = findShortestDistances(start);
    std::vector<std::string> reachable;
    
    for (const auto& pair : distances) {
        if (pair.second != std::numeric_limits<int>::max()) {
            reachable.push_back(pair.first);
        }
    }
    
    return reachable;
}

std::pair<std::vector<std::string>, int> Dijkstra::findShortestPathWithCustomWeight(
    const std::string& start,
    const std::string& end,
    std::function<int(const std::string&, const std::string&)> weightFunction
) {
    // Validate input vertices
    if (!validateVertices(start, end)) {
        return {{}, -1};
    }
    
    // Initialize distance map with infinity
    std::unordered_map<std::string, int> distances;
    std::unordered_map<std::string, std::string> parent;
    std::unordered_set<std::string> visited;
    
    // Get all vertices
    auto vertices = graph.getVertices();
    for (const auto& vertex : vertices) {
        distances[vertex] = std::numeric_limits<int>::max();
    }
    
    // Distance to start vertex is 0
    distances[start] = 0;
    
    // Priority queue for Dijkstra's algorithm
    std::priority_queue<Node, std::vector<Node>, std::greater<Node>> pq;
    pq.push(Node(start, 0));
    
    while (!pq.empty()) {
        Node current = pq.top();
        pq.pop();
        
        std::string currentVertex = current.vertex;
        
        // Skip if already visited
        if (visited.find(currentVertex) != visited.end()) {
            continue;
        }
        
        visited.insert(currentVertex);
        
        // If we reached the end vertex, we can stop early
        if (currentVertex == end) {
            break;
        }
        
        // Get neighbors of current vertex
        auto neighbors = graph.getNeighbors(currentVertex);
        
        for (const auto& neighbor : neighbors) {
            if (visited.find(neighbor) == visited.end()) {
                int edgeWeight = weightFunction(currentVertex, neighbor);
                int newDistance = distances[currentVertex] + edgeWeight;
                
                if (newDistance < distances[neighbor]) {
                    distances[neighbor] = newDistance;
                    parent[neighbor] = currentVertex;
                    pq.push(Node(neighbor, newDistance));
                }
            }
        }
    }
    
    // Check if path exists
    if (distances[end] == std::numeric_limits<int>::max()) {
        return {{}, -1}; // No path found
    }
    
    // Reconstruct path
    std::vector<std::string> path = reconstructPath(parent, start, end);
    
    return {path, distances[end]};
}

std::vector<std::string> Dijkstra::reconstructPath(
    const std::unordered_map<std::string, std::string>& parent,
    const std::string& start,
    const std::string& end
) {
    std::vector<std::string> path;
    
    if (parent.find(end) == parent.end()) {
        return path; // No path found
    }
    
    // Reconstruct path from end to start
    std::string current = end;
    while (current != start) {
        path.push_back(current);
        current = parent.at(current);
    }
    path.push_back(start);
    
    // Reverse to get path from start to end
    std::reverse(path.begin(), path.end());
    
    return path;
}

bool Dijkstra::validateVertices(const std::string& start, const std::string& end) const {
    return graph.hasVertex(start) && graph.hasVertex(end);
}
